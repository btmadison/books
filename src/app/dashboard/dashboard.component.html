<h2 highlight>{{title}}</h2>

<div class="grid grid-pad qa-heroes">
  <dashboard-hero *ngFor="let hero of heroes"  class="col-1-4 qa-hero-list"
    [hero]=hero  (selected)="gotoDetail($event)" >
  </dashboard-hero>
</div>


<!--
Copyright Google LLC. All Rights Reserved.
Use of this source code is governed by an MIT-style license that
can be found in the LICENSE file at https://angular.io/license
-->
