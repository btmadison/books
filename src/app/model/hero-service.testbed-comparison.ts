import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';

import { TestBed } from '@angular/core/testing';
import { HttpResponse } from '@angular/common/http';

import { Hero } from './hero';
import { HeroService } from './hero.service';


// mocking external core services like Http
// to setup create an httpMock with the help of HttpTestingController from HttpClientTestingModule (3.1)
// prep hardcoded data, initiate the client request and setup the assertion that will happen, match the url w/ httpMock.expectOne, using httpMock.expectOne, (3.2)
// use .flush to send the data to the client and .error to emulate error (3.3)
// in error testing, define the error response and bypass the success case (3.4)
// cover the Multiple Request Case, and check the request length (3.5)
// testing PUTs: test the method type and request body that is going out from the client, Use req.event(new HttpResponse({ .. })) to respond' (3.6)

describe('Testing Http with TestBed', () => {
  let heroService: HeroService;
  let httpMock: HttpTestingController;

  // for the hardcoded data that is shared between the tests
  let expectedHeroes: Hero[];
  let updatedHero: Hero;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule // (3.1.1) HttpClientTestingModule in imports
      ],
      providers: [HeroService]
    });

    // (3.1.2) inject the service under test (just like 2.2 )
    heroService = TestBed.inject(HeroService);
    // (3.1.3) mock http using  HttpTestingController
    httpMock = TestBed.inject(HttpTestingController);
    // (service injection similar to other external services, the key difference is the module being imported)
    // (HttpTestingController doesnâ€™t make an HTTP request but allows you to emulate it using hardcoded data.)
  });

  // verify() is used to assert that there are no outstanding http requests
  afterEach(() => httpMock.verify());

  describe('GET requests' , () => {

    beforeEach(() => {
      // (3.2.1) prepare hardcoded data that will be used as the response
      expectedHeroes = [
        { id: 1, name: 'A' },
        { id: 2, name: 'B' },
      ] as Hero[];
    });

    it('(3.2.2) initiate the client request and setup the assertion that will happen,(3.2.3) match the url w/ httpMock.expectOne, (3.3) use .flush to respond to the client ', () => {

      // (3.2.2) initiate the client request, and setup the assertion that will happen once the observable is fulfilled
      heroService.getHeroes().subscribe(
        heroes => expect(heroes).toEqual(expectedHeroes),
        fail // the error case of the observable
      );

      // (3.2.3) expect that a single request has been made which matches the given URL, using httpMock.expectOne
      const req = httpMock.expectOne(heroService.heroesUrl);
      // (3.3) use .flush to respond to the client
      req.flush(expectedHeroes);

      expect(req.request.method).toEqual('GET'); // extra
    });

    it('cover the Empty Response Case', () => {
      heroService.getHeroes().subscribe(
        heroes => expect(heroes.length).toEqual(0),
        fail
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      req.flush([]); // we control the response here
    });

    it('(3.4) in error testing, define the error response and bypass the success', () => {
      const msg = 'Deliberate 404';

      heroService.getHeroes().subscribe(
        heroes => fail('expected to fail'),
        error => expect(error.message).toContain(msg) // 3.4.0 in error testing, define the error response and bypass the success case
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      // 3.4.1 flush with the optional object: status and message
      req.flush(msg, {status: 404, statusText: 'Not Found'});
    });

    it('(3.5) cover the Multiple Request Case, and check the request length', () => {
      heroService.getHeroes().subscribe(); // KEY multiple subscribes
      heroService.getHeroes().subscribe();
      heroService.getHeroes().subscribe(
        heroes => expect(heroes).toEqual(expectedHeroes),
        fail
      );

      const requests = httpMock.match(heroService.heroesUrl);

      // Respond to each request with different mock hero results
      requests[0].flush([]);
      requests[1].flush([{ id: 1, name: 'bob' }]);
      requests[2].flush(expectedHeroes);

      expect(requests.length).toEqual(3); // KEY check request length
    });
  });

  describe('PUT requests', () => {

    beforeEach(() => {
      updatedHero = { id: 1, name: 'A' } as Hero;
    });

    it('(3.6) testing PUTs: test the method type and request body that is going out from the client, Use req.event(new HttpResponse({ .. })) to respond', () => {

      heroService.updateHero(updatedHero).subscribe(
        data => expect(data).toEqual(updatedHero),
        fail
      );

      const req = httpMock.expectOne(heroService.heroesUrl);
      // (3.6.0) KEY with testing PUTs: test the method type and request body that is going out from the client
      expect(req.request.method).toEqual('PUT');
      expect(req.request.body).toEqual(updatedHero);

      // (3.6.1) KEY with PUT: use req.event(new HttpResponse({})) instead of req.flush()
      req.event(
        new HttpResponse({ status: 200, statusText: 'OK', body: updatedHero })
      );
    });

    it('cover the Error Case, same as the GET scenario in (3.4) ', () => {
      const msg = 'Deliberate 404';

      heroService.updateHero(updatedHero).subscribe(
        heroes => fail('expected to fail'),
        error => expect(error.message).toContain(msg) // KEY (same as GET scenario IN 3.4) in error testing, define the error response and bypass the success
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      // KEY: (same as GET scenario) flush with the optional object: status and message
      req.flush(msg, { status: 404, statusText: 'Not Found' });
    });
  });
});
