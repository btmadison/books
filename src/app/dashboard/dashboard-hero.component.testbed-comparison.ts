import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { Hero } from '../model/hero';

import { DashboardHeroComponent } from './dashboard-hero.component';

// unit testing components with @Input and @Output properties
// setup the component with TestBed.configureTestingModule({..}) (satisfy the TS), and create the component with TestBed.createComponent(..) before each test (1.1)
// access the TS with fixture.debugElement.componentInstance / fixture.componentInstance (1.2)
// use fixture.detectChanges() to trigger the change detection (1.3),
// access the template with fixture.debugElement.nativeElement / fixture.debugElement (1.4.1),
// to test nativeElements (ex: @Input) use fixture.debugElement.nativeElement.querySelector('..').textContent (1.4.2)
// to test @Output subscribe to the event emitter and setup what will be emitted (1.5.1),
// access the debugElement with fixture.debugElement.query(By.css('')).nativeElement, trigger event (1.5.2) and verify what is emitted (1.5.3)


describe('Testing Components with TestBed', () => {
describe('unit testing components with @Input and @Output properties', () => {
  let comp: DashboardHeroComponent;
  let fixture: ComponentFixture<DashboardHeroComponent>;

  // (1.1.1) setup the component
  beforeEach(() => {
    TestBed
    .configureTestingModule({
      declarations: [DashboardHeroComponent]
    }); // a few versions ago, we had to wrap this in waitForAsync and use .compileComponents() because the file had an external html and css. Not anymore
  });

  beforeEach(() => {
    // (1.1.2) TestBed.createComponent(): returns a ComponentFixture object which gives access to the TS & the template
    fixture = TestBed.createComponent(DashboardHeroComponent);
    // (1.2) fixture.debugElement.componentInstance / fixture.componentInstance gives access to the TS
    comp = fixture.componentInstance;
  });

  it('(1.1) setup the component, (1.2) access the TS with fixture.componentInstance ' , () => {
    expect(comp).toBeTruthy();
  });

  describe('(1.3) detectChanges() to trigger change detection' , () => {
    let mockHeroInput: Hero;

    beforeEach(() => {
      // this component has an @Input property. So, simulate the @Input hero property being set by the parent
      mockHeroInput = comp.hero =  {id: 42, name: 'new Hero Name'};

      // (1.3) detectChanges() : to update the bindings / trigger change detection
      fixture.detectChanges();
    });

    it('Testing @Input: , (1.4.1) fixture.debugElement.nativeElement to access the template, (1.4.2) fixture.debugElement.nativeElement.querySelector(..).textContent to assert', () => {
      // (1.4.1) fixture.debugElement.nativeElement: gives access to the template native elements.=
      const heroNativeElement = fixture.debugElement.nativeElement;

      // (1.4.2) to test nativeElements use fixture.debugElement.nativeElement.querySelector('..').textContent to assert
      expect(heroNativeElement.querySelector('.hero').textContent).toContain(mockHeroInput.name.toUpperCase());
    });

    it('Testing the @Output: (1.5.1) subscribe to the event emitter and set up what will be emitted, (1.5.2) access the debugElement with fixture.debugElement.query(By.css()).nativeElement and trigger click' , () => {
      let emittedHero: Hero;

      // (1.5.1) subscribe to the event emitter and set up what will be emitted (this.hero will be emitted which we emulate as emittedHero)
      comp.selected.subscribe((hero: Hero) => emittedHero = hero );

      // (1.5.2)  access the debugElement with fixture.debugElement.query(By.css('')).nativeElement and trigger click
      const heroDebugElement =  fixture.debugElement.query(By.css('.hero')).nativeElement;
      heroDebugElement.click();

      // (1.5.3) verify that what is emitted
      expect(emittedHero).toBe(mockHeroInput);
    });
  });

});
});
