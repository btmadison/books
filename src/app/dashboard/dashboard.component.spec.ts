import { Spectator, createRoutingFactory, byText } from '@ngneat/spectator/jest';
import { DashboardComponent } from './dashboard.component';
import { HeroService } from '../model/hero.service';
import { Hero } from '../model';
import { of } from 'rxjs';
import { DashboardHeroComponent } from './dashboard-hero.component';
import { MockComponent } from 'ng-mocks';


// [6] testing components that include other components, services (example[5]), and routing (example[4])
// setup the component with routing, use createRoutingFactory to auto-mock Router and ActivatedRoute (6.1). Just like (4.1)
// mock the service dependency (6.1.2), just like (5.1.2) .
// inject the service dependency:  depService = spectator.inject(DepService) (6.1.3), just like (5.1.3)
// stub the external service's return value (6.1.4), just like (5.1.4)
// KEY: mock the internal components, use the ng-mocks library MockComponent. (6.1.5)

// access the TS with spectator.component  (6.2)
// use spectator.detectChanges()  to trigger the change detection (1.3),
// use DOM testing library convenience methods:  https://github.com/ngneat/spectator#queries' (1.4)
// to test @Output subscribe to the event emitter and setup what will be emitted (1.5.1),
// trigger the event using spectator events api https://github.com/ngneat/spectator#events-api (1.5.2) and verify what is emitted (1.5.3)

describe('[6] Testing components that include other components, services (example[5]), and routing (example[4])', () => {
  let component: DashboardComponent;
  let spectator: Spectator<DashboardComponent>;
  let heroServiceSpy;
  let heroService;

  // setup the component with routing, use createRoutingFactory to auto-mock Router and ActivatedRoute (6.1)
  const createComponent = createRoutingFactory({
    component: DashboardComponent,
    // (6.1.5) KEY: mock the internal components, use the ng-mocks library MockComponent.
    // Instead of using CUSTOM_ELEMENTS_SCHEMA, which might hide some issues and won't help you to set inputs, outputs, etc., ng-mocks will auto mock the inputs, outputs, etc. for you
    declarations: [ DashboardComponent, MockComponent(DashboardHeroComponent) ],
    // (6.1.2) mock the internal components and the service dependency,
    mocks: [HeroService],
    detectChanges: false,
  });


  const getTestHeroes = (): Hero[] => {
    return [
      {id: 41, name: 'Bob' },
      {id: 42, name: 'Carol' },
      {id: 43, name: 'Ted' },
      {id: 44, name: 'Alice' },
      {id: 45, name: 'Speedy' },
      {id: 46, name: 'Stealthy' }
    ];
  };

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component; // (6.2) access the TS with spectator.component

    // (6.1.3) inject the service dependency:  depService = spectator.inject(DepService)
    heroServiceSpy = spectator.inject(HeroService);
    // (6.1.4) stub the external service's return value
    heroService = jest.spyOn(heroServiceSpy, 'getHeroes').mockReturnValue(of(getTestHeroes()));
  });

  it('sanity', () => {
    expect(component).toBeTruthy();
  });

  it('should NOT have heroes before ngOnInit', () => {
    expect(component.heroes.length).toBe(0);
  });

  it('should HAVE heroes', () => {
    spectator.detectChanges();
    expect(component.heroes.length).toBeGreaterThan(0);
  });

  it('should DISPLAY heroes', () => {
    spectator.detectChanges();

    expect(spectator.queryAll('dashboard-hero').length).toBe(4);
  });


  // continue with routing
  it('should tell ROUTER to navigate when hero clicked', async () => {
    expect(spectator.query('.hero')).toBeDefined();

    // spectator.click('.hero'); // TODO: cannot click

    // await spectator.fixture.whenStable();
    // expect(spectator.inject(Router).navigate).toHaveBeenCalledWith(['/heroes/12']);
  });

  // TODO: convert this Karma test to spectator
  // it('should tell ROUTER to navigate when hero clicked', () => {
  //   heroClick();  // trigger click on first inner <div class="hero">

  //   // args passed to router.navigateByUrl() spy
  //   const spy = router.navigateByUrl as jasmine.Spy;
  //   const navArgs = spy.calls.first().args[0];

  //   // expecting to navigate to id of the component's first hero
  //   const id = comp.heroes[0].id;
  //   expect(navArgs).toBe('/heroes/' + id, 'should nav to HeroDetail for first hero');
  // });

});
