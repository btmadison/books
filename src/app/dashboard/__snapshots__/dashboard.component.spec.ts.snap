// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[6] Testing components that include other components, services (example[5]), and routing (example[4]) should HAVE heroes after ngOnInit, snapshot test the component after onInit 1`] = `
<app-dashboard
  __ngContext__={[Function LRootView]}
  heroService={[Function HeroService]}
  heroes={[Function Array]}
  router={[Function Object]}
>
  <h2
    highlight=""
  >
    Top 4 Heroes
  </h2><div
    class="grid grid-pad qa-heroes"
  >
    <dashboard-hero
      class="col-1-4 qa-hero-list-item"
    />
    <dashboard-hero
      class="col-1-4 qa-hero-list-item"
    />
    <dashboard-hero
      class="col-1-4 qa-hero-list-item"
    />
    <dashboard-hero
      class="col-1-4 qa-hero-list-item"
    />
    
  </div>
</app-dashboard>
`;

exports[`[6] Testing components that include other components, services (example[5]), and routing (example[4]) should NOT have heroes before ngOnInit, snapshot test before onInit 1`] = `
<app-dashboard
  __ngContext__={[Function LRootView]}
  heroService={[Function HeroService]}
  heroes={[Function Array]}
  router={[Function Object]}
>
  <h2
    highlight=""
  >
    
  </h2><div
    class="grid grid-pad qa-heroes"
  >
    
  </div>
</app-dashboard>
`;
